courseworks/week8/yushengzhao/unitdisk.jl

```diff
@@ -0,0 +1,1340 @@
+### A Pluto.jl notebook ###
+# v0.19.22
+
+using Markdown
+using InteractiveUtils
+
+# ╔═╡ 454da7d4-d103-11ed-389e-df4bbc1c34ff
+begin
+    using Graphs,Plots, GraphRecipes, Optim, Optimisers, Random, ForwardDiff, Combinatorics
+end
+
+# ╔═╡ c82fd2da-71cb-47ec-a487-af8100bdb705
+md"
+# ChaptGPT's definition of Unit Disk
+
+A unit disk embedding of a graph is a way of representing a graph in the plane
+such that each vertex is mapped to a point in the plane and each edge is
+represented by a curve connecting the corresponding points, such that the length
+of each edge is at most 1 (i.e., the points representing the endpoints of the
+edge are at most distance 1 apart). This type of embedding is often used in
+wireless networking and sensor placement applications, where the nodes in the
+graph represent wireless devices or sensors and the edges represent
+communication or sensing ranges.
+"
+
+# ╔═╡ 5e0f25a5-096a-4702-86d2-a472e901d743
+#this is used everywhere, DRY
+global E = [(1, 2), (1, 3),
+	(2, 3), (2, 4), (2, 5), (2, 6),
+	(3, 5), (3, 6), (3, 7),
+	(4, 5), (4, 8),
+	(5, 6), (5, 8), (5, 9),
+	(6, 7), (6, 8), (6, 9),
+	(7,9), (8, 9), (8, 10), (9, 10)]
+
+# ╔═╡ 3bee9c9a-3802-4180-b5d8-240bef0f1237
+begin
+edges = Edge.(E)
+# create a graph from the vertex and edge lists above
+g = SimpleGraphFromIterator(edges)
+end
+
+# ╔═╡ 70091c2b-2f0e-4767-a1d8-ab23ee947de8
+graphplot(g)
+
+# ╔═╡ 91bf072a-df45-494a-b393-46fa311c057f
+function dist(a::AbstractVector{T},b::AbstractVector{T}) where T
+    return sqrt((a[1]-b[1])^2 + (a[2]-b[2])^2)
+end
+
+# ╔═╡ 67822069-6128-4600-9e8a-84e3a1e75b16
+function loss_f(x::AbstractMatrix)
+    loss = 0.0
+    for (a,b) in combinations(1:10,2)
+		if (a,b) in E
+			prop_dist = dist(x[:,a],x[:,b])
+		else
+			prop_dist = 1/dist(x[:,a],x[:,b])
+		end
+
+		if prop_dist > 1
+			loss  +=  prop_dist
+		end
+
+    end
+    return loss
+end
```

Comment:
To get a better convergent rate, you may set ```prop_dist > 1``` as ```prop_dist > 0.95``` so that the result will be more robust.